lib/kubernetes-deploy/resource_deployer.rb

```diff
@@ -0,0 +1,236 @@
+    end
+    measure_method(:deploy_all_resources, 'normal_resources.duration')
+
+    def deploy_resources(resources, prune: false, verify:, record_summary: true)
+      return if resources.empty?
+      deploy_started_at = Time.now.utc
+
+      if resources.length > 1
+        logger.info("Deploying resources:")
+        resources.each do |r|
+          logger.info("- #{r.id} (#{r.pretty_timeout_type})")
+        end
+      else
+        resource = resources.first
+        logger.info("Deploying #{resource.id} (#{resource.pretty_timeout_type})")
+      end
+
+      # Apply can be done in one large batch, the rest have to be done individually
+      applyables, individuals = resources.partition { |r| r.deploy_method == :apply }
+      # Prunable resources should also applied so that they can  be pruned
+      pruneable_types = @prune_whitelist.map { |t| t.split("/").last }
+      applyables += individuals.select { |r| pruneable_types.include?(r.type) }
+
+      individuals.each do |r|
+        r.deploy_started_at = Time.now.utc
+        case r.deploy_method
+        when :replace
+          _, _, replace_st = kubectl.run("replace", "-f", r.file_path, log_failure: false)
+        when :replace_force
+          _, _, replace_st = kubectl.run("replace", "--force", "--cascade", "-f", r.file_path,
+            log_failure: false)
+        else
+          # Fail Fast! This is a programmer mistake.
+          raise ArgumentError, "Unexpected deploy method! (#{r.deploy_method.inspect})"
+        end
+
+        next if replace_st.success?
+        # it doesn't exist so we can't replace it
+        _, err, create_st = kubectl.run("create", "-f", r.file_path, log_failure: false)
+
+        next if create_st.success?
+        raise FatalDeploymentError, <<~MSG
+          Failed to replace or create resource: #{r.id}
+          #{err}
+        MSG
+      end
+
+      apply_all(applyables, prune)
+
+      if verify
+        watcher = KubernetesDeploy::ResourceWatcher.new(resources: resources, deploy_started_at: deploy_started_at,
+          timeout: @max_watch_seconds, task_config: @task_config, sha: @current_sha)
+        watcher.run(record_summary: record_summary)
+      end
+    end
+
+    def apply_all(resources, prune)
+      return unless resources.present?
+      command = %w(apply)
+
+      Dir.mktmpdir do |tmp_dir|
+        resources.each do |r|
+          FileUtils.symlink(r.file_path, tmp_dir)
+          r.deploy_started_at = Time.now.utc
+        end
+        command.push("-f", tmp_dir)
+
+        if prune
+          command.push("--prune")
+          if @selector
+            command.push("--selector", @selector.to_s)
+          else
+            command.push("--all")
+          end
+          @prune_whitelist.each { |type| command.push("--prune-whitelist=#{type}") }
+        end
+
+        output_is_sensitive = resources.any?(&:sensitive_template_content?)
+        global_mode = resources.all?(&:global?)
```

Comment:
So in the mixed-mode scenario currently possible under the old deploy task, we pass a namespace arg that is just ignored by kubectl as appropriate? I guess that's what we've always done, and it works. 🤔